opal/corelib/string.rb

```diff
@@ -1017,7 +1017,7 @@ def scan(pattern, &block)
         }
       }
 
-      #{$~ = `match_data`}
+      #{$~ = `match_data` unless no_matchdata}
```

Comment:
Since it's a private API I think we can expect it to be either `true` or `false`, and move the if to JS:

```js
if (!no_matchdata) #{$~ = `match_data`}
```